### ISSUE TITLE
Enhancement: add session.IsUpgrade / setupEventArgs.IsUpgrade

### ISSUE THREAD

**Issue by manytwo** [oleg-shilo/wixsharp#31]
There is a property for all cases:

- IsInstalling   
- IsUpgrading    
- IsModifying   
- IsUninstalling 
- IsInstalled   

But not when an installing product is an upgrade, we can see only during uninstall (IsUpgrading ).

May be:
```CS
public bool IsUpgrade { get { return Data["UPGRADEFOUND"].IsNotEmpty(); } }
```

It is a very minor thing and it can be done easily manually (session.Property("UPGRADEFOUND")), but it should be added to complete the pattern.

**Comment by oleg-shilo** [comment]
"Upgrading" is a tricky state to detect. 
MSI architecture and state management made it really convoluted and this resulted in the limitations of how MSI reports the runtime state. The limitations are captured in the header of this file: https://github.com/oleg-shilo/wixsharp/blob/master/Source/src/WixSharp.UI/ManagedUI/MsiRuntime.cs
All `SetupEventArgs` state properties are based on these properties.

The unpleasant truth is that one cannot just check the built-in properties and reliably conclude that it is an upgrade as the desired properties combination is not present at startup.

One possible work around is to have a custom property and set it to the result of "is_prev_version_present" analysis.  The older `Upgrade` WiX element (Wix#: `Project.MajorUpgradeStrategy`) supports setting the detected version value to the custom property. And WixSharp creates such a property (UPGRADEFOUND) so indeed you can use it. 

However it is not a reliable approach as WiX recommend another approach `MajorUpgrade` element (Wix#: `Project.MajorUpgrade`). An this element does not support tunneling the analysis result to the custom properties. Thus UPGRADEFOUND is never set. So... unfortunate "no" for this one.

Alternative work around would be to do custom detection by yourself ihe BeforeInstall event.  You do have AppSearch:
```C#
project.GUID = new Guid("6f330b47-2577-43ad-9095-1861ba25889b");
    ...
static void project_BeforeInstall(SetupEventArgs e)
{
    Version prev_ver  = AppSearch.GetProductVersion("{6f330b47-2577-43ad-9095-1861ca31dfd4}");
```

I will keep this issue open as want to see if this analysis can be done by Wix# for ManagedProjects.

**Comment by oleg-shilo** [comment]
Just tested my suggestion and it will not work. 

The problem is that you cannot call `AppSearch.GetProductVersion` as you don't know the ProductCode to pass in. It is different for any new version. You need to use UpgradeCode which is shared by all versions of the same product. And it cannot be used with `GetProductVersion`. You need to enumerate through all so called "related products" and then probe their ProductCode. Yes MSI is that twisted! :(

I have created a utility method that encapsulates this:
```C#
static void project_BeforeInstall(SetupEventArgs e)
{
    var installedVersion = AppSearch.GetProductVersionFromUpgradeCode(e.UpgradeCode);
}
```
The code is committed and just waiting for the release.

**Comment by manytwo** [comment]
Nice, I didn't see this msi api wrapper, I was using Microsoft.Deployment.WindowsInstaller.  I didn't have to embed it  now.
I was pretty sure that "UPGRADEFOUND" was a standard msi property like "UPGRADINGPRODUCTCODE", I was persuaded to seen it in msi doc (maybe a monday morning without coffe haha). 

When I was testing Microsoft.Deployment.WindowsInstaller.GetRelatedProducts, I remember it found 2 related products during uninstall in upgrade process, at BeforeInstall step. Don't now if it matter for your code.

**Comment by oleg-shilo** [comment]
OK, you might be right. I just found with google some of UPGRADEFOUND's traces on WiX. But not on MSI. So it looks like it's a custom property which is "standard" for WiX. But MSI doco still has nothing on it. At least I couldn't find any.  

Anyway, even if UPGRADEFOUND is a standard prop but it is not a part of "MajorUpgrade" element workflow. So it still cannot be used.

As for that wrapper, I just committed it tonight so no wonder you didn't find it :)
I modified MajorUpgrade samples and now they show detected installed version.
